File: content/problems/2035-partition-array-into-two-arrays-to-minimize-sum-difference/solutions/meet-in-the-middle-by-cardinality/Solution.kt
class Solution {
    fun minimumDifference(nums:IntArray):Int{
        val n=nums.size/2;
        val left=Array(n+1){
            mutableListOf<Long>()
        };
        val right=Array(n+1){
            mutableListOf<Long>()
        };
        val total=nums.sumOf{
            it.toLong()
        };
        for(mask in 0 until (1 shl n)){
            var a=0L;
            var b=0L;
            val count=Integer.bitCount(mask);
            for(bit in 0 until n)if(mask and (1 shl bit)!=0){
                a+=nums[bit];
                b+=nums[n+bit]
            };
            left[count].add(a);
            right[count].add(b)
        };
        right.forEach{
            it.sort()
        };
        var answer=Long.MAX_VALUE;
        for(count in 0..n)for(a in left[count]){
            val values=right[n-count];
            val target=total/2-a;
            var index=values.binarySearch(target);
            if(index<0)index=-index-1;
            for(position in maxOf(0,index-1)..minOf(values.lastIndex,index+1))answer=minOf(answer,kotlin.math.abs(total-2*(a+values[position])))
        };
        return answer.toInt()
    }
}

File: content/problems/2035-partition-array-into-two-arrays-to-minimize-sum-difference/solutions/meet-in-the-middle-by-cardinality/solution.cpp
class Solution {
    public: int minimumDifference(vector<int>& nums){
        int n=nums.size()/2;
        vector<vector<long long>> left(n+1),right(n+1);
        long long total=accumulate(nums.begin(),nums.end(),0LL);
        for(int mask=0;mask<(1<<n);++mask){
            long long a=0,b=0;
            int count=__builtin_popcount((unsigned)mask);
            for(int bit=0;bit<n;++bit)if(mask&(1<<bit)){
                a+=nums[bit];
                b+=nums[n+bit];
            }
            left[count].push_back(a);
            right[count].push_back(b);
        }
        for(auto& values:right)sort(values.begin(),values.end());
        long long answer=LLONG_MAX;
        for(int count=0;count<=n;++count)for(long long a:left[count]){
            auto& values=right[n-count];
            long long target=total/2-a;
            auto iterator=lower_bound(values.begin(),values.end(),target);
            if(iterator!=values.end())answer=min(answer,llabs(total-2*(a+*iterator)));
            if(iterator!=values.begin()){
                --iterator;
                answer=min(answer,llabs(total-2*(a+*iterator)));
            }
        }
        return answer;
    }
};
